crates/uv-distribution/src/metadata/requires_dist.rs

```diff
@@ -260,6 +251,92 @@ fn apply_source_strategy(
+        >,
+        groups: &BTreeMap<&GroupName, &Vec<DependencyGroupSpecifier>>,
+        name: &GroupName,
+        parents: &BTreeSet<GroupName>,
+    ) -> Result<Vec<uv_pep508::Requirement<VerbatimParsedUrl>>, MetadataError> {
+        let Some(specifiers) = groups.get(name) else {
+            // Missing group
+            let parent_name = parents
+                .iter()
+                .last()
+                .expect("We should always have a parent when a group is missing");
+            return Err(MetadataError::GroupNotFound(
+                name.clone(),
+                parent_name.clone(),
+            ));
+        };
+
+        if parents.contains(name) {
+            // A cycle was detected
+            // TODO(zanieb): We should probably add an error variant for this, we probably want to
+            // show the full `parents` list to demonstrate the cycle?
+            return Ok(Vec::new());
+        }
+
+        if let Some(result) = resolved.get(name) {
+            // Already resolved this group
+            // TODO(zanieb): Can we avoid cloning here? Will we want to clone anyway?
+            return Ok(result.as_ref().unwrap().clone());
+        }
+
+        let mut parents = parents.clone();
+        parents.insert(name.clone());
+
+        let mut requirements = Vec::new();
+        for specifier in *specifiers {
+            match specifier {
+                DependencyGroupSpecifier::Requirement(requirement) => {
+                    match uv_pep508::Requirement::<VerbatimParsedUrl>::from_str(requirement) {
+                        Ok(requirement) => requirements.push(requirement),
+                        Err(err) => {
+                            return Err(MetadataError::GroupParseError(
+                                name.clone(),
+                                requirement.clone(),
+                                Box::new(err),
+                            ));
+                        }
+                    }
+                }
+                DependencyGroupSpecifier::Table {
+                    include_group: Some(include_name),
+                } => {
+                    let include_name = GroupName::from_str(include_name).unwrap();
+                    let include_requirements =
+                        resolve_group(resolved, groups, &include_name, &parents)?;
+                    requirements.extend(include_requirements);
+                }
+                DependencyGroupSpecifier::Table {
+                    include_group: None,
+                } => {}
+            }
+        }
+
+        resolved.insert(name.clone(), Ok(requirements.clone()));
+        Ok(requirements)
+    }
+
+    let mut resolved = BTreeMap::new();
+
+    for name in groups.keys() {
+        // TODO(zanieb): Raise errors? or defer them until a group is needed per the PEP's
+        // suggestion?
+        let _ = resolve_group(&mut resolved, groups, name, &BTreeSet::new());
```

Comment:
I would say just raise here.

Reply:
It defies the "SHOULD" recommendation in the PEP, but we will generally need to resolve all groups anyway for locking so I think that's reasonable for now.